Shared spack (#47615)
spack/spack: lib/spack/docs/configuration.rst

#. Each must be preceded with the ``--config-scope`` or ``-C`` flag.
#. They must be ordered from lowest to highest precedence.

.. _scopes-and-xdg-compliance:

^^^^^^^^^^^^^^^^^^^^^^^^^
Scopes and XDG Compliance
^^^^^^^^^^^^^^^^^^^^^^^^^

Spack respects XDG variables, and the search location for user configuration files can be affected by them.
When ``XDG_CONFIG_HOME`` is not defined, Spack assumes the XDG default value of ``~/.config``, and will apply user and this-spack configuration files located in ``~/.config/spack``.
Defining ``XDG_CONFIG_HOME`` will change where Spack searches for configuration files.

> To override this behavior, define ``SPACK_USER_CONFIG_PATH`` to be the desired path.

Reviewer: ```suggestion
To override this behavior, define ``SPACK_CONFIG_HOME`` to be the desired path.
```
We should update this to match the rest of the variables, and only maintain `SPACK_USER_CONFIG_PATH` as an alias for backwards compatibility.

Author: Apologies but `SPACK_CONFIG_HOME` no longer exists and I think it would actually be misleading for it to exist (would imply there's a similar fall-through path for config as for data/cache/state). I'll be trying to update the docs more, but this is sufficiently-in-flux enough that at the moment my focus is keeping the PR description as definitive, and then updating docs if we agree as part of preliminary review.

There is often a need for platform-specific configuration settings.
For example, on most platforms, GCC is the preferred compiler.
However, on macOS (darwin), Clang often works for more packages, and is set as the default compiler.

> This configuration is set in ``$(prefix)/etc/spack/defaults/darwin/packages.yaml``, which is included as by ``$(prefix)/etc/spack/defaults/include.yaml``.

Reviewer: ```suggestion
This configuration is set in ``$(prefix)/etc/spack/defaults/darwin/packages.yaml``, which is included by ``$(prefix)/etc/spack/defaults/include.yaml``.
```

The corresponding variables that describe where this data is placed are:

* ``$data_home``
* ``$cache_home``
* ``$state_home``
* ``$user_cache_path``: legacy variable, equivalent to ``$state_home``

You can refer to these variables when configuring locations for stages, misc cache, etc.
Furthermore each is controlled by a fall-through scheme.
For example ``$data_home`` evaluates to one of the following (highest-priority first):

#. ``SPACK_DATA_HOME`` env var if that is set
#. ``SPACK_HOME`` env var
#. ``config:locations:data``
#. ``config:locations:home``
#. ``XDG_DATA_HOME`` env var if that is set
#. Under the default for ``XDG_DATA_HOME``: ``~/.local/share/spack``

Of particular interest is where the environments and installs are placed by Spack, because these can take up a lot of space.
Generally speaking these are controlled by ``$data_home``.
Older installs of spack placed these within ``$spack``, and fallback scheme in these cases is augmented to prefer these old locations if no data is detected in the corresponding new locations:

* ``$default_install_root``: the location where installs go by default.
  Overridden by ``config:install_tree:root``.

>   Prefers ``$data_home/installs``, but if there are no installs there but there are installs in the old location within ``$spack``, then the old location will be used.

Reviewer: ```suggestion
  Prefers ``$data_home/installs``, but if there are no installs there and there are installs in the old location ``$spack/opt/spack``, then the old location will be used.
```

>   Prefers ``$data_home/installs``, but if there are no installs there but there are installs in the old location ``$spack/opt/spack``, then the old location will be used.

Reviewer: ```suggestion
  Prefers ``$data_home/installs``, but if there are no installs there and there are installs in the old location ``$spack/opt/spack``, then the old location will be used.
```

* ``$default_envs_root``: the location where environments are managed by default.
  Overridden by ``config:environments_root``.

>   Prefers ``$data_home/envs`` but if there are no envs there and there are envs in the old location in ``$spack``, then the old location will be used.

Reviewer: ```suggestion
  Prefers ``$data_home/envs`` but if there are no envs there and there are envs in the old location ``$spack/var/spack/environments``, then the old location will be used.
```

.. code-block:: console

  $ export SPACK_DISABLE_LOCAL_CONFIG=true
  $ export SPACK_USER_CACHE_PATH=/tmp/spack

.. _xdg_overrides:

Overriding default paths with XDG variables
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

While Spack will by default use locations within the ``$spack`` and the user's home directory, it can now store all state in locations goverened by XDG variables.
Many default storage locations have changed in 1.0, so this section will summarize the changes, and how XDG variables can help users organize their Spack artifacts.

``XDG_DATA_HOME`` is used to store long-lived data.
Its default value is ``$HOME/.local/share``.
The following items are always stored by default using ``$XDG_DATA_HOME``:

* Modules: ``$XDG_DATA_HOME/spack/``
* Package indices: ``$XDG_DATA_HOME/$spack_instance_id/spack``

The data listed below will be placed in ``$spack`` by default.
However, if ``XDG_DATA_HOME`` is set, they will be stored under that path.
The default values for these are as follows:

* Source caches: ``$spack/var/spack/cache`` or ``$XDG_DATA_HOME/spack/downloads``.
* The install tree: ``$spack/opt/spack`` or ``$XDG_DATA_HOME/spack/installs``.

> * Environment management: ``$spack/var/spack/environtments`` or ``$XDG_DATA_HOME/spack/environments``

Reviewer: We need to update these to not default into the Spack instance (unless the new location is unused and the old one is used, for backwards compatibility).

``XDG_STATE_HOME`` is used to store data that is useful if persistent, but not integral to Spack's functionality.
If not defined, its default value is ``~/.local/state``.
``misc_cache`` is placed by default using this variable.

* ``misc_cache``: ``$XDG_STATE_HOME/$spack_instance_id/spack``

``XDG_CACHE_HOME`` is used to store temporary data.
If not defined, its default value is ``~/.cache``.
Build stages are placed by default using this variable.

* Build stages: ``$XDG_CACHE_HOME/spack``

The user configuration scope's files will be located with ``XDG_CONFIG_HOME``.
Its default value is ``~/.config``.

* User configuration scope: ``$XDG_CONFIG_HOME/spack``

Spack also includes the variables ``SPACK_DATA_HOME``, ``SPACK_CONFIG_HOME``, and ``SPACK_STATE_HOME`` that map directly to the XDG variables described above.
They work the same way, but have higher precedence than the XDG variables.
If the Spack-specific environment variables *are not* defined, Spack will uses the XDG variables with a suffix of "/spack" to define ``$spack_state_home``, ``$spack_data_home``, and ``spack_cache_home``.

> If they are defined, they are used directly without any additional suffix.

Reviewer: We should rephrase this to make the initial statement not seem to contradict the suffixing clarification later. Maybe something like

```suggestion
Spack also includes the variables ``SPACK_DATA_HOME``, ``SPACK_CONFIG_HOME``, and ``SPACK_STATE_HOME`` that override the XDG variables described above.
They work the same way, but have higher precedence than the XDG variables.
If the Spack-specific environment variables *are not* defined, Spack will uses the XDG variables with a suffix of "/spack" to define ``$spack_state_home``, ``$spack_data_home``, and ``spack_cache_home``.
If they are defined, they are used directly without any additional suffix.
```

Author: I rewrote this section (among other things the rewrite reflects that config file vars like `$spack_state_home` no longer exist in this PR): `configuration.rst` - "Spack-specific variables controlling data location"
